### PR TITLE
Lua object class support

### PR DESCRIPTION
**Pull request by noahdesu**

### CODE REVIEW

**Comment by yuyuyu101** [comment]
COOL!

Not sure we need to contains lua source in ceph repo. Specify a external directory may better?

**Comment by cxwshawn** [comment]
cool!  looking forward to being merged !

**Comment by liewegas** [comment]
Does it makes sense to use a submodule instead of an extracted tarball for lua?

**Comment by noahdesu** [comment]
@yuyuyu101 @liewegas I have no issue with Lua being in a submodule (what's the motivation?) The only important thing is that we control that submodule because we may want to customize the VM in the future, but I'm assuming I target a ceph/lua repo?

**Comment by liewegas** [comment]
yeah.  the submodule makes it easier to move between lua versions, keeps 
lua and ceph code separate, and avoids munging up any source code stats.

**Comment by noahdesu** [comment]
sounds good. i'll need to have someone create ceph/lua repo.. looks like I don't have permission.

**Comment by liewegas** [comment]
Forked!  https://github.com/ceph/lua

**Comment by noahdesu** [comment]
Thanks. I need push access to that repo. It isn't a vanilla Lua distribution, and its build configuration has been modified to assume cmake, so we'll need to hack it a bit to play nice with automake too.

**Comment by batrick** [comment]
Hi @noahdesu, this is an exciting feature! I'm still looking through the code and have yet to try this out but I was wondering why you include Roberto's Lua struct module when Lua 5.3 now has a [pack](http://www.lua.org/manual/5.3/manual.html#pdf-string.pack) function (which is loosely based on struct).

**Comment by noahdesu** [comment]
Hey @batrick thanks for pointing that out! I didn't realize that Lua 5.3 had this function. I wanted to include some useful serialization modules out-of-the-box so people could immediately start using this. It sounds like it might be OK to nuke the struct module for now. Let me know if you have any other questions or issues using cls_lua.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -1188,6 +1188,48 @@ if(${WITH_RADOSGW})
...
+    lua/lua-5.3.2/src/lvm.c
+    lua/lua-5.3.2/src/lzio.c
+)
+
+add_library(cephlua SHARED ${liblua_srcs})
+
+target_include_directories(cephlua
+    PUBLIC lua/lua-5.3.2/src/
+)
+
```
Comment (cbodley) [inline src/CMakeLists.txt]: could we move the cmake bits into a lua/CMakeLists.txt please?

Comment (noahdesu) [inline src/CMakeLists.txt]: Absolutely. I'm in the process of stuffing Lua into a git submodule and I'll add this to the todo list.

**Comment by noahdesu** [comment]
@liewegas mind adding me to ceph/lua so I can push there?

**Comment by yehudasa** [comment]
@noahdesu I updated the permissions there

**Comment by noahdesu** [comment]
Thanks @yehudasa that worked. @liewegas ok got it all submodule-ified. @cbodley could you take a look at the new cmake files?

**Comment by cbodley** [comment]
@noahdesu the cmake looks good, thanks

**Comment by noahdesu** [comment]
Rebased to remove merge recent merge conflicts.

File: src/cls/lua/cls_lua.cc [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+
+/*
+ * cls_cxx_map_get_val
+ */
+static int clslua_map_get_val(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  const char *key = luaL_checkstring(L, 1);
+  bufferlist  *bl = clslua_pushbufferlist(L, NULL);
+  int ret = cls_cxx_map_get_val(hctx, key, bl);
```
Comment (batrick) [inline src/cls/lua/cls_lua.cc]: I believe `ret` needs to be negated on error right? In my tests, it returns a negative value which gives us:

`Unknown error -2`

in the logs.

Comment (noahdesu) [inline src/cls/lua/cls_lua.cc]: Do you mean negated to resolve to the correct errno string via strerror, or negated to change the logic?

Comment (batrick) [inline src/cls/lua/cls_lua.cc]: To get the correct errno string I believe.

Comment (noahdesu) [inline src/cls/lua/cls_lua.cc]: Thanks for catching this Patrick. I pushed a fix for it that seems to work (note that I also rebased to fix a merge conflict in master) https://github.com/ceph/ceph/commit/50739d704b5ee9a3858a8c39c98faa1a6659833d

**Comment by athanatos** [comment]
There was some discussion of this at one of the recent CDM's.  I think one of the main concerns was that we need a mechanism to whitelist classes for users?

**Comment by noahdesu** [comment]
@athanatos yes. i just created PR #9972 for review, that adds the loading and caps whitelist changes for object classes.

**Comment by noahdesu** [comment]
@athanatos not sure what the next steps are for this. the whitelisting feature is merged now which was the latest blocker.

**Comment by athanatos** [comment]
This looks right to me.  I think it should go through the rados suite.

**Comment by athanatos** [comment]
retest this please

**Comment by yuriw** [comment]
@noahdesu this PR failed the build, pls fix and tag `needs-qa`

http://gitbuilder.sepia.ceph.com/gitbuilder-ceph-rpm-centos7-amd64-basic/log.cgi?log=cd3d9f482a50fc6d6cb0c7e0ac4a23ca42d97f45

**Comment by noahdesu** [comment]
@yuriw packages are building on centos-7 now and can be re-tested. thanks.

File: src/cls/CMakeLists.txt [hunk 3]
```diff
@@ -135,3 +135,17 @@ endif (WITH_CEPHFS)
...
 add_library(cls_numops_client STATIC numops/cls_numops_client.cc)
 
+add_library(cls_lua SHARED
+    lua/cls_lua.cc
+    lua/lua_bufferlist.cc
+)
+set_target_properties(cls_lua PROPERTIES VERSION "1.0.0" SOVERSION "1")
+install(TARGETS cls_lua DESTINATION ${cls_dir})
+target_link_libraries(cls_lua
+    liblua
```
Comment (dmick) [inline src/cls/CMakeLists.txt]: I suspect this should just be "lua"; the built package is trying to find a dependency "libliblua"

**Comment by yuriw** [comment]
@noahdesu from IRC

(12:56:22 PM) yuriw: rados jobs failing ! http://tracker.ceph.com/issues/16860
(12:57:26 PM) dmick: it seems to be that the key error is
(12:57:38 PM) dmick: Error: Package: 1:ceph-base-11.0.0-1084.gf2349d2.x86_64 (Ceph) 
(12:57:38 PM) dmick: Requires: libliblua.so()(64bit)
(12:58:53 PM) yuriw: libliblua is a submodule I needed for one of PR
(01:00:08 PM) dmick: submodule?  not in master
(01:00:37 PM) dmick: but this is claiming that ceph-base was built with a package dependency on it
(01:03:28 PM) dmick: I don't see anything in that PR that changes ceph.spec though ?
(01:03:46 PM) dmick: https://github.com/ceph/ceph/pull/7338
(01:07:48 PM) dmick: I don't know how that package dependency got in there, but I'm guessing rpmbuild inferred it, and libliblua isn't being packaged.  Since it says "libliblua", I suspect it's an error in the CMake stuff, where the library name shouldn't have 'lib' on it, maybe?.... in any event you should drop that PR and get Noah to investigate
(01:10:29 PM) dmick: yuriw: https://github.com/ceph/ceph/pull/7338/files#r72853201
(01:12:30 PM) zackc: wusui: ok, cool

Pls investigate and re-tag as needed.

File: src/cls/CMakeLists.txt (hunk 3 above)
Comment (cbodley) [inline src/cls/CMakeLists.txt]: from https://github.com/ceph/lua/blob/lua-5.1/CMakeLists.txt, the project actually defines a 'liblua' target that we refer to here:
```cmake
add_library ( liblua SHARED ${SRC_LIB} ${LUA_DLL_RC} ${LUA_DEF} )
target_link_libraries ( liblua ${LIBS} )
set_target_properties ( liblua PROPERTIES OUTPUT_NAME lua CLEAN_DIRECT_OUTPUT 1 )
```
(the `OUTPUT_NAME lua` property is what prevents it from naming the output file libliblua.so)

Comment (cbodley) [inline src/cls/CMakeLists.txt]: oops, i see that the lua-5.3-ceph branch https://github.com/ceph/lua/blob/lua-5.3-ceph/CMakeLists.txt comments out the `OUTPUT_NAME` part. wonder if that's related

**Comment by noahdesu** [comment]
@yuriw gosh, ok thanks. sorry about these bugs I never expected this sort of thing since liblua isn't actually being installed :/

**Comment by yuriw** [comment]
@noahdesu np :) if you have a quick fix I can take it back for re-test (if it's today)

**Comment by noahdesu** [comment]
Ah, shoot, I won't be able to get to this until later this weekend.

**Comment by yuriw** [comment]
@noahdesu np we will take it in next batch then

File: src/cls/CMakeLists.txt (hunk 3 above)
Comment (noahdesu) [inline src/cls/CMakeLists.txt]: Hmm, I'll try this naming change, but I don't see anything in rpmbuild docs that suggest the naming causes the auto dependency. What I'm wondering is if since liblua is being built as a shared library, but not being installed, that the dependency between the object class and liblua is being recorded. with a vstart setup this doesn't show up. linking statically into the object class may work.

**Comment by noahdesu** [comment]
@yuriw @cbodley @dmick so i think i've resolved this by building a static lua library (i don't see the auto dependencies being created), but I am having an issue that is preventing me from running tests: on the trust debian builder all of the ceph_test_xxx have this warning:

    warning: dpkg-shlibdeps: couldn't find library libgmock_main.so needed by debian/ceph-test/usr/bin/ceph_test_cls_rbd (ELF format: 'elf64-x86-64'; RPATH: '/usr/bin/../lib')

the full list of these warnings is at the bottom of http://gitbuilder.sepia.ceph.com/gitbuilder-ceph-deb-trusty-amd64-basic/log.cgi?log=abc439fc14fe963f1f239804ca1dca85cb7e48d1

so when I install the ceph-test package I cannot run the tests. I rebased on master to see if I needed to pick up any changes, but it isn't clear how / if I am introducing this. any thoughts?

**Comment by ktdreyer** [comment]
Noah, could you build against the system lua instead of bundling it into Ceph? (for example RHEL/CentOS 7 have lua-devel-5.1.4-14.el7)

When we bundle lua's source code into Ceph, the Ceph build time becomes longer, and the Ceph project becomes responsible for fixing CVEs in the bundled version, etc.

**Comment by batrick** [comment]
Does RHEL7 have Lua 5.3? Optionally building against the system Lua 5.3 sounds reasonable to me.

Also, Lua's source is pretty small. On my humble laptop a parallel make builds in 2.5 seconds.

**Comment by noahdesu** [comment]
I'm not opposed to linking against a system library, but it seems the precedent is to avoid this when embedding the VM mostly due to compatibility and customization issues. For instance Redis disables byte-code execution with a patch to the VM source, and other things are possible like switching out the native numeric data type.

From the Lua documentation: "Different versions are really different. The API is likely to be a little different (but with compatibility switches), and there is no ABI compatibility: applications that embed Lua and C libraries for Lua must be recompiled. The virtual machine is also very likely to be different in a new version: Lua programs that have been precompiled for one version will not load in a different version."

That might be manageable when creating releases, but strikes me as a headache waiting to happen especially for people mixing Lua versions. Thoughts?

**Comment by ktdreyer** [comment]
There are pros and cons to bundling, and I wanted to at least have the discussion so I understand why we're doing it in each case. As long as the team is ok maintaining it when security fixes come up, I'm ok with it.

**Comment by noahdesu** [comment]
Sounds good. I'm happy to do any tests or research on the issue for everyone to agree on an approach.

**Comment by noahdesu** [comment]
the libgmock_main.so problem was caused by some cmake stuff that the lua submodule was doing that was causing various other libraries to be built as shared libraries rather than the expected static variety.

@yuriw anyway, let's give this a shot again. i was able to install from packages on deb/rhel and things are working now.

**Comment by tchaikov** [comment]
> the libgmock_main.so problem was caused by some cmake stuff that the lua submodule was doing that was causing various other libraries to be built as shared libraries rather than the expected static variety.

@noahdesu in case you didn't find it, see https://github.com/ceph/lua/blob/lua-5.1/cmake/dist.cmake#L77 . i suffered from this when working with gmock also.

**Comment by noahdesu** [comment]
@tchaikov nice thanks for the reference, that was a tough one to track down. is there any reason to patch the lua variable? when i was reading through dist.cmake i didn't see anything in there we care about. it was packaging and installation stuff.

**Comment by tchaikov** [comment]
Noah, see https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html . this variable changes the default behavior of `add_library()`.

**Comment by noahdesu** [comment]
@tchaikov cool thanks. i'll get that fixed up tomorrow.

**Comment by noahdesu** [comment]
@tchaikov I think I misread your comment. I understand the global impact of the `BUILD_SHARED_LIBS` variable. I was asking about patching it indirectly: that variable is set in `lua/cmake/dist.cmake` which contains stuff related to installation and packaging---cmake boilerplate that we do not depend on, including shared library building which doesn't matter since we build lua as  a static library. So I disabled the entire inclusion of `lua/cmake/dist.cmake`, including the `BUILD_SHARED_LIBS` global library. This file also sets numerous cmake variables, so we don't need to worry about them clobbering other settings either.

**Comment by batrick** [comment]
@noahdesu, have you seen the thread on ceph-devel concerning Lua packaging? I'd link to it but gmane appears down (maybe permanently).

In particular, I'm concerned about including Ceph libraries in the lua submodule. Can those be moved `src/cls/lua`?

**Comment by noahdesu** [comment]
@batrick thanks for the pointer, i'll chime in on the list. about your question here: you are referring to cmsgpack and the bufferlist bindings? absolutely, they can be moved to src/cls/lua. lua was originally moved to a submodule and the reason people wanted that was that it 1) screwed up LOC stats and 2) ostensibly was easier to manage. Since I didn't cmsgpack it seemed more appropriate there, but its easy to move. The reasoning behind leaving bufferlist bindings there was because it was general purpose.

**Comment by tchaikov** [comment]
> I was asking about patching it indirectly: that variable is set in `lua/cmake/dist.cmake` which contains stuff related to installation and packaging---cmake boilerplate that we do not depend on, including shared library building which doesn't matter since we build lua as  a static library. So I disabled the entire inclusion of `lua/cmake/dist.cmake`, including the `BUILD_SHARED_LIBS` global variable.

@noahdesu sorry, i failed to understand your question. ic. i thought a less intrusive way would be easier for maintenance. but when i am looking at https://github.com/ceph/lua/commits/lua-5.3-ceph: well, we have already added a handful of patches on top of lua-5.3 branch, yet another one won't hurt. =)

**Comment by michaelsevilla** [comment]
What is the status on this?

**Comment by noahdesu** [comment]
I need to take another look at the thread on ceph-devel to make sure nothing is missing, but I think we just need to slap needs-qa back on this.

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2016-09-19_19:42:27-rados-wip-yuri-testing2_2016_09_19-distro-basic-smithi/

rerun http://pulpito.ceph.com/yuriw-2016-09-20_14:50:21-rados-wip-yuri-testing2_2016_09_19-distro-basic-smithi/
